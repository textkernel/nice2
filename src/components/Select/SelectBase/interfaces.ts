import { GetItemPropsOptions, GetToggleButtonPropsOptions } from 'downshift';
import { Props as ListRendererProps } from '../SuggestionsList/SuggestionsList';

export interface CommonProps<S> extends React.HTMLAttributes<HTMLDivElement> {
    /** an array of objects that will be used to render the suggestions list. */
    suggestions: S[];
    /** suggestionToString(suggestion) should return a string to be displayed in the UI. e.g.: suggestion => suggestion.name */
    suggestionToString: (suggestions?: S | null) => string;
    /** input field ref */
    inputRef?: React.RefObject<HTMLInputElement>;
    /** root wrapper ref */
    rootRef?: React.RefObject<HTMLDivElement>;
    /** suggestions list ref */
    listRef?: React.RefObject<HTMLUListElement>;
    /** onBlur() is called when the component is blurred */
    onBlur?: () => void;
    /** onSelectionChange() called when a suggestion is selected or removed */
    onSelectionChange: (item: S) => void;
    /** onInputValueChange(inputValue) called when the input values is changed. Can be used to implement the component as controlled component */
    onInputValueChange: (value: string) => void;
}

export interface CommonPropsWithClear<S> extends CommonProps<S> {
    /** to be shown as clear button title */
    clearTitle?: string;
    /** function to be called if FieldWrapper clear button is clicked */
    onClearAllSelected?: () => void;
    /** show Clear button on hover even if there are no selectedSuggestions passed */
    showClearButton?: boolean;
}

export interface Props<S> extends CommonPropsWithClear<S> {
    keepExpandedAfterSelection?: boolean;
    /** will be called when list of suggestions should be rendered */
    listRenderer: (props: Omit<ListRendererProps<S>, 'noSuggestionsPlaceholder'>) => ReactNode;
    /** a function that renders the top part of the component when it is focused  */
    focusedRenderer: (helpers: {
        getInputProps: (options: GetItemPropsOptions<S>) => object;
        getToggleButtonProps: (options: GetToggleButtonPropsOptions) => object;
        onBlur: () => void;
        inputValue: string;
    }) => ReactNode;
    /** a function that renders the top part of the component when it is blurred  */
    blurredRenderer: (helpers: {
        getInputProps: (options: GetItemPropsOptions<S>) => object;
        getToggleButtonProps: (options: GetToggleButtonPropsOptions) => object;
    }) => ReactNode;
}
