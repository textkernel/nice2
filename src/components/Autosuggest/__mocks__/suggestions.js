export const SUGGESTIONS = [
    { name: 'Sun' },
    { name: 'Moon' },
    { name: 'Mercury' },
    { name: 'Venus' },
    { name: 'Earth' },
    { name: 'Mars' },
    { name: 'Jupiter' },
    { name: 'Saturn' },
    { name: 'Neptun' }
];

export const SUGGESTION_TO_STRING = item => (item ? item.name : '');
