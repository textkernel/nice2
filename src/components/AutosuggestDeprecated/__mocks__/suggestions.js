export const SUGGESTIONS = [
    { name: 'Sun' },
    { name: 'Moon' },
    { name: 'Mercury' },
    { name: 'Venus' },
    { name: 'Earth' },
    { name: 'Mars' },
    { name: 'Jupiter' },
    { name: 'Saturn' },
    { name: 'Neptun' },
];

export const COMPLEX_SUGGESTIONS = [
    { name: 'Sun', type: 'Star', count: 1234 },
    { name: 'Moon', type: 'Star', count: 1234 },
    { name: 'Mercury', type: 'Planet', count: 1234 },
    { name: 'Venus', type: 'Planet', count: 1234 },
    { name: 'Earth', type: 'Planet', count: 1234 },
    { name: 'Mars', type: 'Planet', count: 1234 },
    { name: 'Jupiter', type: 'Planet', count: 1234 },
    { name: 'Saturn', type: 'Planet', count: 1234 },
    { name: 'Neptun', type: 'Planet', count: 1234 },
];

export const SUGGESTION_TO_STRING = (item) => (item ? item.name : '');
