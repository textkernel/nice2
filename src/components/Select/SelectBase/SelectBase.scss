.SelectBase {
    position: relative;
    min-height: 38px;

    &__main {
        position: relative;
        box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.1);
        background-color: var(--color-background);
        transition-property: box-shadow, transform, border-color;
        transition-duration: var(--transition-duration);
        border-radius: var(--border-radius);

        &--focused {
            position: absolute;
            z-index: 400;
            width: 100%;
        }

        &--isProminent {
            transform-origin: left top;

            &.SelectBase__main--focused {
                transform: scale(1.1);
                box-shadow: 0 5px 15px 0 rgba(0, 0, 0, 0.4);
            }
        }
    }

    &__field {
        padding: 0;
    }

    &__spacedElem {
        margin: var(--spacing-normal);
    }

    &__list--focused {
        max-height: 214px;
        overflow: auto;
        border-top: 1px solid var(--border-color-light);
        margin-top: -1px;
    }

    &__loaderContainer {
        width: 250px;
    }
}
