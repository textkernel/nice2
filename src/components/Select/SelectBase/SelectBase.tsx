import * as React from 'react';
import Downshift from 'downshift';
import { bem } from '../../../utils/bem';
import { FieldWrapper } from '../../FieldWrapper';
import { List } from '../../List';
import { ENTER_KEY } from '../../../constants';
import { Props } from './interfaces';
import styles from './SelectBase.scss';

const { block, elem } = bem('SelectBase', styles);

export function SelectBase<S>(props: Props<S>) {
    const {
        suggestions,
        suggestionToString,
        clearTitle,
        showClearButton,
        onBlur,
        onSelectionChange,
        onInputValueChange,
        onClearAllSelected,
        inputRef: inputRefFromProps,
        rootRef: rootRefFromProps,
        listRef: listRefFromProps,
        listRenderer,
        focusedRenderer,
        blurredRenderer,
        keepExpandedAfterSelection,
        ...rest
    } = props;

    const [inputRef, setInputRef] = React.useState(
        inputRefFromProps || React.createRef<HTMLInputElement>()
    );
    const [rootRef, setRootRef] = React.useState(
        rootRefFromProps || React.createRef<HTMLDivElement>()
    );
    const [listRef, setListRef] = React.useState(
        listRefFromProps || React.createRef<HTMLUListElement>()
    );

    const [inputValue, setInputValue] = React.useState('');
    const [inputValueRecall, setInputValueRecall] = React.useState('');
    const [focused, setFocused] = React.useState(false);

    // focus input field if component is focused
    React.useEffect(() => {
        if (focused) {
            inputRef.current?.focus();
        }
    }, [focused, inputRef]);

    // update inputRef from props if changed
    React.useEffect(() => {
        if (inputRefFromProps) {
            setInputRef(inputRefFromProps);
        }
    }, [inputRefFromProps]);

    // update listRef from props if changed
    React.useEffect(() => {
        if (listRefFromProps) {
            setListRef(listRefFromProps);
        }
    }, [listRefFromProps]);

    // update rootRef from props if changed
    React.useEffect(() => {
        if (rootRefFromProps) {
            setRootRef(rootRefFromProps);
        }
    }, [rootRefFromProps]);

    const handleBlur = () => {
        setInputValue('');
        setInputValueRecall('');
        onBlur?.();
    };

    const handleChange = (selectedItem, downshift) => {
        const { clearSelection, openMenu } = downshift;

        setInputValue('');
        clearSelection();

        if (selectedItem) {
            onSelectionChange(selectedItem);
        }

        if (!keepExpandedAfterSelection) {
            handleBlur();
        } else {
            openMenu();
        }
    };

    const handleInputValueChange = (inputVal, changes) => {
        if (changes.type === Downshift.stateChangeTypes.changeInput) {
            setInputValue(inputVal);
            setInputValueRecall(inputVal);
            if (listRef.current) {
                listRef.current.scrollTop = 0;
            }

            onInputValueChange?.(inputVal);
        }
    };

    const handleClearSelectedSuggestions = e => {
        e.stopPropagation();
        onClearAllSelected?.();
    };

    const focus = openMenu => {
        openMenu();
        setFocused(true);
    };

    const handleWrapperClick = openMenu => () => {
        if (!focused) {
            focus(openMenu);
        }
    };

    const handleWrapperKeyDown = openMenu => e => {
        if (!focused && e.key === ENTER_KEY) {
            focus(openMenu);
        }
    };

    const stateReducer = (state, changes) => {
        switch (changes.type) {
            case Downshift.stateChangeTypes.clickItem:
            case Downshift.stateChangeTypes.keyDownEnter:
                return {
                    ...changes,
                    highlightedIndex: state.highlightedIndex,
                    isOpen: keepExpandedAfterSelection,
                };
            default:
                return changes;
        }
    };

    const stateUpdater = (change, state) => {
        if (state.isOpen !== focused) {
            setFocused(state.isOpen);
        }
    };

    const stateAndProps = { props, focused };

    const getInputPropsWithUpdatedRef = getInputProps => inputProps => {
        return inputRefFromProps
            ? getInputProps(inputProps)
            : getInputProps({ ...inputProps, ref: inputRef });
    };

    return (
        <div {...rest} ref={rootRef} {...block(stateAndProps)}>
            <Downshift
                onChange={handleChange}
                itemToString={suggestionToString}
                onOuterClick={handleBlur}
                stateReducer={stateReducer}
                onStateChange={stateUpdater}
                onInputValueChange={handleInputValueChange}
                inputValue={inputValue}
                defaultHighlightedIndex={0}
            >
                {({
                    getInputProps,
                    getMenuProps,
                    getItemProps,
                    getToggleButtonProps,
                    highlightedIndex,
                    openMenu,
                }) => (
                    <div {...elem('main', stateAndProps)}>
                        <FieldWrapper
                            clearLabel={clearTitle}
                            onClear={handleClearSelectedSuggestions}
                            showClearButton={!focused && showClearButton}
                            isFocused={focused}
                            onClick={handleWrapperClick(openMenu)}
                            onKeyDown={handleWrapperKeyDown(openMenu)}
                            {...elem('field', stateAndProps)}
                        >
                            {focused
                                ? focusedRenderer({
                                      getInputProps: getInputPropsWithUpdatedRef(getInputProps),
                                      getToggleButtonProps,
                                      onBlur: handleBlur,
                                      inputValue,
                                  })
                                : blurredRenderer({ getInputProps, getToggleButtonProps })}
                            <List
                                {...getMenuProps({
                                    ...elem('list', stateAndProps),
                                    ref: listRef.current,
                                    isControlledNavigation: true,
                                })}
                            >
                                {focused
                                    ? listRenderer({
                                          suggestionToString,
                                          suggestions,
                                          getItemProps,
                                          highlightedIndex,
                                          inputValue: inputValueRecall,
                                      })
                                    : null}
                            </List>
                        </FieldWrapper>
                    </div>
                )}
            </Downshift>
        </div>
    );
}

SelectBase.defaultProps = {
    showClearButton: false,
    keepExpandedAfterSelection: false,
    clearTitle: '',
};

SelectBase.displayName = 'SelectBase';
