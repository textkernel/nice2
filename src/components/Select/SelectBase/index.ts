export { SelectBase } from './SelectBase';
export { CommonProps, CommonPropsWithClear, Props } from './interfaces';
