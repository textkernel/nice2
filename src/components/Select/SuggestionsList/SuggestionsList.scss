.SuggestionsList {
    &__optimizerWrapper {
        height: 214px;
    }
}
