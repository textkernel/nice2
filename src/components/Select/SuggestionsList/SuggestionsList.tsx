import * as React from 'react';
import { GetItemPropsOptions } from 'downshift';
import { bem } from '../../../utils/bem';
import { Text, MarkedText } from '../../Text';
import { ListOptimizer, ListItem } from '../../List';
import styles from './SuggestionsList.scss';

const { elem } = bem('SuggestionsList', styles);

export interface Props<S> {
    /** An array of objects that will be used to render the suggestions list. */
    suggestions: S[];
    /** Enable ListOptimizer component for decreasing render time */
    useOptimizeRender?: boolean;
    /** suggestionToString(suggestion) should return a string to be displayed in the UI. e.g.: suggestion => suggestion.name */
    suggestionToString: (suggestions: S) => string;
    /** to be shown when no suggestions are available */
    noSuggestionsPlaceholder?: string;
    /** a function which gets props for the item in the list */
    getItemProps: (options: GetItemPropsOptions<S>) => object;
    /** index of the item from the list to be highlighted */
    highlightedIndex: number | null;
    /** input field value to be highlighted in the item from the list */
    inputValue: string;
}

export function SuggestionsList<S>(props: Props<S>) {
    const {
        suggestionToString,
        useOptimizeRender,
        suggestions,
        noSuggestionsPlaceholder,
        getItemProps,
        highlightedIndex,
        inputValue,
    } = props;

    if (noSuggestionsPlaceholder && (!suggestions || !suggestions.length)) {
        return (
            <ListItem disabled>
                <Text context="muted">{noSuggestionsPlaceholder}</Text>
            </ListItem>
        );
    }

    // eslint-disable-next-line react/display-name
    const renderItem = ({ key, index, style = {} }) => (
        <ListItem
            key={key}
            style={style}
            {...getItemProps({
                item: suggestions[index],
                index,
            })}
            isHighlighted={highlightedIndex === index}
            highlightContext="brand"
        >
            <MarkedText marker={inputValue} inline>
                {suggestionToString(suggestions[index])}
            </MarkedText>
        </ListItem>
    );

    return (
        <>
            {useOptimizeRender ? (
                <div {...elem('optimizerWrapper')}>
                    <ListOptimizer rowCount={suggestions.length}>{renderItem}</ListOptimizer>
                </div>
            ) : (
                <>
                    {suggestions.map((item, index) => {
                        const key = suggestionToString(item);
                        return renderItem({ key, index });
                    })}
                </>
            )}
        </>
    );
}

SuggestionsList.defaultProps = {
    noSuggestionsPlaceholder: '',
    useOptimizeRender: false,
};

SuggestionsList.displayName = 'SuggestionsList';
