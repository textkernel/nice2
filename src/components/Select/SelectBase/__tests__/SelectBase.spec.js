import React from 'react';
import toJson from 'enzyme-to-json';
import { SelectBase } from '../SelectBase';
import { SuggestionsList } from '../../SuggestionsList';
import { SUGGESTIONS, SUGGESTION_TO_STRING } from '../../../Autosuggest/__mocks__/suggestions';

describe('SelectBase', () => {
    const inputRef = React.createRef();
    const suggestions = SUGGESTIONS;
    const suggestionToString = SUGGESTION_TO_STRING;
    const noSuggestionsPlaceholder = 'No suggestions...';
    const mockOnSelectionChange = jest.fn();
    const mockOnInputValueChange = jest.fn();
    const mockOnClearAllSelected = jest.fn();
    const mockOnBlur = jest.fn();
    const mockRender = jest.fn(({ getInputProps }) => <input {...getInputProps()} />);
    const mockRenderWithRef = jest.fn(({ getInputProps }) => (
        <input ref={inputRef} {...getInputProps()} />
    ));

    let wrapper;
    let inputNode;

    const setFocusOnInput = () =>
        wrapper
            .find('.SelectBase__field')
            .at(1)
            .simulate('click');

    beforeEach(() => {
        wrapper = mount(
            <SelectBase
                suggestions={suggestions}
                suggestionToString={suggestionToString}
                noSuggestionsPlaceholder={noSuggestionsPlaceholder}
                listRenderer={listProps => <SuggestionsList {...listProps} />}
                focusedRenderer={mockRender}
                blurredRenderer={mockRender}
                onSelectionChange={mockOnSelectionChange}
                onInputValueChange={mockOnInputValueChange}
                onClearAllSelected={mockOnClearAllSelected}
                onBlur={mockOnBlur}
            />
        );
        inputNode = wrapper.find('input').getDOMNode();
    });

    describe('rendering', () => {
        it('should initially render empty component correctly', () => {
            expect(toJson(wrapper)).toMatchSnapshot();
        });
        it('should render empty component correctly when focused', () => {
            setFocusOnInput();
            wrapper.setProps({ suggestions: [] });
            expect(wrapper.find('li')).toHaveLength(0);
            expect(inputNode).toBe(document.activeElement);
        });
        it('should render noSuggestions placeholder when empty suggestions list is passed', () => {
            wrapper.setProps({ suggestions: [] });
            wrapper.find('input').simulate('change', { target: { value: 'driver' } });
            setFocusOnInput();

            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('li')).toHaveLength(1);
            expect(
                wrapper
                    .find('li')
                    .childAt(0)
                    .text()
            ).toEqual(noSuggestionsPlaceholder);
        });
        it('should render all suggestions from the list', () => {
            setFocusOnInput();

            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('li')).toHaveLength(suggestions.length);
        });
        it('should use alternative list render function if passed', () => {
            const listRendererMock = jest.fn(() => <li>My list</li>);

            wrapper.setProps({ listRenderer: listRendererMock });
            setFocusOnInput();

            expect(listRendererMock).toHaveBeenCalledWith(
                expect.objectContaining({
                    getItemProps: expect.any(Function),
                    inputValue: expect.any(String),
                    noSuggestionsPlaceholder: expect.any(String),
                    highlightedIndex: expect.any(Number),
                    suggestionToString: expect.any(Function),
                    suggestions: expect.any(Array),
                })
            );
            expect(wrapper.find('li')).toHaveLength(1);
        });
        it('should render Clear button when showClearButton is set to true', () => {
            const clearTitle = 'Clear';
            wrapper.setProps({ clearTitle, showClearButton: true });

            setFocusOnInput();
            wrapper
                .find('li')
                .first()
                .children()
                .simulate('click');
            wrapper.find('input').simulate('blur');

            expect(wrapper.find('Button').text()).toBe(clearTitle);
            expect(wrapper.find('li')).toHaveLength(0);
        });
    });
    describe('focusing and blurring the search field', () => {
        it('should set focus when wrapper element is clicked', () => {
            const focusSpy = jest.spyOn(inputNode, 'focus');

            expect(inputNode).not.toBe(document.activeElement);
            expect(focusSpy).not.toHaveBeenCalled();

            setFocusOnInput();

            expect(inputNode).toBe(document.activeElement);
            expect(focusSpy).toHaveBeenCalled();
        });
        it('should implement direct control functionality with a ref passed from the outside', () => {
            wrapper.setProps({
                ref: inputRef,
                focusedRenderer: mockRenderWithRef,
                blurredRenderer: mockRenderWithRef,
            });
            wrapper.update();
            const focusSpy = jest.spyOn(inputNode, 'focus');

            expect(inputNode).not.toBe(document.activeElement);
            expect(focusSpy).not.toHaveBeenCalled();

            setFocusOnInput();
            wrapper
                .find('li')
                .first()
                .children()
                .simulate('click');

            expect(inputNode).toBe(document.activeElement);
            expect(focusSpy).toHaveBeenCalled();

            wrapper.find('input').simulate('blur');

            expect(mockOnBlur).toHaveBeenCalled();
        });
    });
    describe('callbacks', () => {
        describe('onSelectionChange', () => {
            it('should be called on clicking on a suggestion', () => {
                setFocusOnInput();

                expect(mockOnSelectionChange).not.toHaveBeenCalled();

                wrapper
                    .find('li')
                    .first()
                    .children()
                    .simulate('click');

                expect(mockOnSelectionChange).toHaveBeenCalled();
            });
        });
        it('should call onClearAllSelected on Clear button click', () => {
            const clearTitle = 'Clear';
            wrapper.setProps({ clearTitle, showClearButton: true });

            setFocusOnInput();
            wrapper
                .find('li')
                .first()
                .children()
                .simulate('click');
            wrapper.find('input').simulate('blur');

            wrapper.find('Button').simulate('click');
            expect(mockOnClearAllSelected).toHaveBeenCalled();
        });
        it('should call onBlur', () => {
            setFocusOnInput();
            wrapper
                .find('li')
                .first()
                .children()
                .simulate('click');
            wrapper.find('input').simulate('blur');

            expect(mockOnBlur).toHaveBeenCalled();
        });
        it('should lose focus when suggestion is selected', () => {
            expect(inputNode).not.toBe(document.activeElement);

            setFocusOnInput();

            wrapper
                .find('li')
                .at(0)
                .children()
                .simulate('click');

            expect(wrapper.find('li')).toHaveLength(0);
        });
        it('should stay focused when suggestion is selected with keepExpandedAfterSelection set to true', () => {
            wrapper.setProps({ keepExpandedAfterSelection: true });
            expect(inputNode).not.toBe(document.activeElement);

            setFocusOnInput();
            expect(inputNode).toBe(document.activeElement);

            wrapper
                .find('li')
                .at(0)
                .childAt(0)
                .simulate('click');

            expect(inputNode).toBe(document.activeElement);
            expect(wrapper.find('li')).toHaveLength(suggestions.length);
        });
        it('should call onInputValueChange when typing into input field', () => {
            expect(mockOnInputValueChange).not.toHaveBeenCalled();

            wrapper.find('input').simulate('change', { target: { value: 'driver' } });

            expect(mockOnInputValueChange).toHaveBeenCalled();
        });
        it('should not clear the input field when a suggestion was selected', () => {
            const textInputValue = 'driver';
            wrapper.find('input').simulate('change', { target: { value: textInputValue } });

            expect(wrapper.find('input').props().value).toEqual(textInputValue);

            wrapper
                .find('li')
                .first()
                .children()
                .simulate('click');

            expect(wrapper.find('input').props().value).toEqual('');
        });
    });
});
