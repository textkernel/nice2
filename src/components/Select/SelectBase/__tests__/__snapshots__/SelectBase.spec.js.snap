// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectBase rendering should initially render empty component correctly 1`] = `
<SelectBase
  blurredRenderer={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "getInputProps": [Function],
            "getToggleButtonProps": [Function],
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": <input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-controls={null}
            aria-labelledby="downshift-0-label"
            autoComplete="off"
            id="downshift-0-input"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyDown={[Function]}
            value=""
          />,
        },
      ],
    }
  }
  clearTitle=""
  focusedRenderer={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "getInputProps": [Function],
            "getToggleButtonProps": [Function],
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": <input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-controls={null}
            aria-labelledby="downshift-0-label"
            autoComplete="off"
            id="downshift-0-input"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyDown={[Function]}
            value=""
          />,
        },
      ],
    }
  }
  keepExpandedAfterSelection={false}
  listRenderer={[Function]}
  noSuggestionsPlaceholder="No suggestions..."
  onBlur={[MockFunction]}
  onClearAllSelected={[MockFunction]}
  onInputValueChange={[MockFunction]}
  onSelectionChange={[MockFunction]}
  showClearButton={false}
  suggestionToString={[Function]}
  suggestions={
    Array [
      Object {
        "name": "Sun",
      },
      Object {
        "name": "Moon",
      },
      Object {
        "name": "Mercury",
      },
      Object {
        "name": "Venus",
      },
      Object {
        "name": "Earth",
      },
      Object {
        "name": "Mars",
      },
      Object {
        "name": "Jupiter",
      },
      Object {
        "name": "Saturn",
      },
      Object {
        "name": "Neptun",
      },
    ]
  }
>
  <div
    className="SelectBase"
    noSuggestionsPlaceholder="No suggestions..."
  >
    <Downshift
      defaultHighlightedIndex={0}
      defaultIsOpen={false}
      environment={[Window]}
      getA11yStatusMessage={[Function]}
      inputValue=""
      itemToString={[Function]}
      onChange={[Function]}
      onInputValueChange={[Function]}
      onOuterClick={[Function]}
      onSelect={[Function]}
      onStateChange={[Function]}
      onUserAction={[Function]}
      scrollIntoView={[Function]}
      selectedItemChanged={[Function]}
      stateReducer={[Function]}
      suppressRefError={false}
    >
      <div
        aria-expanded={false}
        aria-haspopup="listbox"
        aria-labelledby="downshift-0-label"
        aria-owns={null}
        className="SelectBase__main"
        role="combobox"
      >
        <FieldWrapper
          className="SelectBase__field"
          clearLabel=""
          isFocused={false}
          onClear={[Function]}
          onClick={[Function]}
          onKeyDown={[Function]}
          showClearButton={false}
        >
          <div
            className="FieldWrapper SelectBase__field"
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            <div
              className="FieldWrapper__content"
            >
              <input
                aria-activedescendant={null}
                aria-autocomplete="list"
                aria-controls={null}
                aria-labelledby="downshift-0-label"
                autoComplete="off"
                id="downshift-0-input"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyDown={[Function]}
                value=""
              />
              <List
                aria-labelledby="downshift-0-label"
                doSelectOnNavigate={false}
                id="downshift-0-menu"
                isControlledNavigation={true}
                isDivided={false}
                role="listbox"
              >
                <ul
                  aria-labelledby="downshift-0-label"
                  className="List"
                  id="downshift-0-menu"
                  role="listbox"
                />
              </List>
            </div>
          </div>
        </FieldWrapper>
      </div>
    </Downshift>
  </div>
</SelectBase>
`;
